### PR TITLE
Diffusivity docs

### PR DESCRIPTION
**Pull request by navidcy**

### CODE REVIEW

File: src/TurbulenceClosures/discrete_diffusion_function.jl [hunk 1]
```diff
@@ -5,46 +5,68 @@ using Oceananigans.Utils: instantiate
...
-```
-with parameters
-```
-func(i, j, k, grid, clock, model_fields, parameters)
-```
-"""
-struct DiscreteDiffusionFunction{LX, LY, LZ, P, F} 
-    func :: F
-    parameters :: P
+  If `LX == LY == LZ == nothing` the function call requires locations in the signature.
```
Comment (simone-silvestri) [inline src/TurbulenceClosures/discrete_diffusion_function.jl:32]:
```suggestion
  If `LX == LY == LZ == nothing` the diffusivity is evaluated at the required locations and the function call requires locations `ℓx, ℓy, ℓz` in the signature.
```

File: src/TurbulenceClosures/discrete_diffusion_function.jl [hunk 2]
```diff
@@ -5,46 +5,68 @@ using Oceananigans.Utils: instantiate
...
+
+  - With `parameters` is not `nothing`, `func` is called with the signature
+
+    ```julia
+    func(i, j, k, grid, ℓx, ℓy, ℓz, clock, model_fields, parameters)
+    ```
+
+  **With locations**
+
+  If `LX, LY, LZ != (nothing, nothing, nothing)` the function call *does not require*
```
Comment (simone-silvestri) [inline src/TurbulenceClosures/discrete_diffusion_function.jl:54]:
```suggestion
  If `LX, LY, LZ != (nothing, nothing, nothing)` the diffusivity is evaluated at `(LX, LY, LZ)` and interpolated onto the required locations. In this case, the function call *does not require*
```

File: src/TurbulenceClosures/discrete_diffusion_function.jl [hunk 3]
```diff
@@ -5,46 +5,68 @@ using Oceananigans.Utils: instantiate
...
+  locations in the signature and the output is automatically interpolated on the correct
+  location. In this case, the diffusivity `func` is called with the signature:
```
Comment (simone-silvestri) [inline src/TurbulenceClosures/discrete_diffusion_function.jl:56]:
```suggestion
  locations in the signature. The diffusivity `func` is called with the signature:
```

File: src/TurbulenceClosures/discrete_diffusion_function.jl [hunk 4]
```diff
@@ -5,46 +5,68 @@ using Oceananigans.Utils: instantiate
...
+
+  1. Without parameters:
+
+    ```julia
+    func(i, j, k, grid, clock, model_fields)
+    ```
+
+  2. With `parameters` is not `nothing`, `func` is called with the signature
```
Comment (simone-silvestri) [inline src/TurbulenceClosures/discrete_diffusion_function.jl:64]:
```suggestion
  2. When `parameters` is not `nothing`, `func` is called with the signature
```

File: src/TurbulenceClosures/discrete_diffusion_function.jl [hunk 5]
```diff
@@ -5,46 +5,68 @@ using Oceananigans.Utils: instantiate
...
+    ```julia
+    func(i, j, k, grid, ℓx, ℓy, ℓz, clock, model_fields)
+    ```
+
+    where `i, j, k` are the indices, `grid` is `model.grid`, `ℓx, ℓy, ℓz` are the
+    instantiated versions of `LX, LY, LZ`, `clock.time` is the current simulation time,
+    `clock.iteration` is the current model iteration, and `model_fields` is a
+    `NamedTuple` with `u, v, w`, the fields in `model.tracers` and the `model.auxiliary_fields`.
+
+  - With `parameters` is not `nothing`, `func` is called with the signature
```
Comment (simone-silvestri) [inline src/TurbulenceClosures/discrete_diffusion_function.jl:46]:
```suggestion
  - When `parameters` is not `nothing`, `func` is called with the signature
```

**Review by simone-silvestri — APPROVED** [review]

File: src/TurbulenceClosures/discrete_diffusion_function.jl [hunk 6]
```diff
@@ -5,46 +5,67 @@ using Oceananigans.Utils: instantiate
...
-with parameters
-```
-func(i, j, k, grid, clock, model_fields, parameters)
-```
-"""
-struct DiscreteDiffusionFunction{LX, LY, LZ, P, F} 
-    func :: F
-    parameters :: P
+  If `LX == LY == LZ == nothing` the diffusivity is evaluated at the required locations. In this case, the function call *requires passing* locations `ℓx, ℓy, ℓz` in the signature.
+  In this case, the diffusivity `func` is called with the signature:
```
Comment (glwagner) [inline src/TurbulenceClosures/discrete_diffusion_function.jl:33]:
```suggestion
  If `LX == LY == LZ == nothing` diffusivity `func` is called with the signature:
```

This confused me a little bit. Is this the default? I think it's important to concisely describe what the user input needs to be. The explanation that "the diffusivity is evaluated at the required locations" doesn't help I don't think.

Comment (glwagner) [reply src/TurbulenceClosures/discrete_diffusion_function.jl:33]: Once the user input is described, one could add a follow-up sentence like "the purpose of this feature is..."

Comment (navidcy) [reply src/TurbulenceClosures/discrete_diffusion_function.jl:33]: So there is no default for the `DiscreteDiffusionFunction`; seems like it's a required kwarg.

But various closures have defaults and I added remarks on that in those docs; see [8ce5499](https://github.com/CliMA/Oceananigans.jl/pull/3133/commits/8ce5499e2df2af29c80644b1b3e0e40e6a1b9d03).

File: src/TurbulenceClosures/turbulence_closure_implementations/scalar_diffusivity.jl [hunk 7]
```diff
@@ -91,15 +91,13 @@ ScalarDiffusivity{ExplicitTimeDiscretization}(ν=0.0, κ=Oceananigans.Turbulence
...
-julia> parameters = (depth_scale = 120.0,)
-
 julia> @inline function κ(i, j, k, grid, p)
            z = znode(i, j, k, grid)
            return 2000 * exp(z / p.depth_scale)
        end
 κ (generic function with 1 method)
 
-julia> ScalarDiffusivity(κ = κ, discrete_form = true, loc = (Center, Center, Face), parameters = parameters)
+julia> ScalarDiffusivity(κ = κ, discrete_form = true, loc = (Center, Center, Face), parameters = (depth_scale = 120.0,))
```
Comment (glwagner) [inline src/TurbulenceClosures/turbulence_closure_implementations/scalar_diffusivity.jl:100]:
```suggestion
julia> ScalarDiffusivity(κ = κ, discrete_form = true, loc = (Center, Center, Face), parameters = (; depth_scale = 120.0))
```

File: src/TurbulenceClosures/discrete_diffusion_function.jl (hunk 6 above)
Comment (navidcy) [reply src/TurbulenceClosures/discrete_diffusion_function.jl:33]: @glwagner, are you happy with how the docstrings of ScalarDiffusivity/ScalarBiharmonicDiffusivity explain the situation?

The `DiscreteDiffusionFunction` docs are not user-facing (they are not even included in the Docs/Library). Happy to make them better tho.

Comment (glwagner) [reply src/TurbulenceClosures/discrete_diffusion_function.jl:33]: I think like DiscreteForcing and DiscreteBoundaryFunction, DiscreteDiffusionFunction is not user-facing as @simone-silvestri says. Users build this object through ScalarDiffusivity right?

Comment (glwagner) [reply src/TurbulenceClosures/discrete_diffusion_function.jl:33]: Well, maybe I was confused. I was just looking at the edits in the PR.

Comment (glwagner) [reply src/TurbulenceClosures/discrete_diffusion_function.jl:33]: I think if its not user facing the docs should actually say that. Also the docs should be oriented towards developers, not users.

Comment (glwagner) [reply src/TurbulenceClosures/discrete_diffusion_function.jl:33]: In summary I am fine for the updates proposed on these developer-facing docs. I suspect there are some possible improvements of the user API; as the discussion reveals there are still some wrinkles

Comment (navidcy) [reply src/TurbulenceClosures/discrete_diffusion_function.jl:33]: Not all of the changes in this PR was about developer-facing docs though.

**Pull request merged**